### PR TITLE
[TF FE] Set single output tensor name for TF1 formats.

### PR DESCRIPTION
**Pull request by popovaan**
Description:
Set single output tensor name for TF1 formats, the format is "`<op name>:<output port index>`".

Ticket: 118376

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names
* [x]  Transformation preserves tensor names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests
* [x]  Transformation tests
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list
* [x]  Guide on how to convert the **public** model
* [x]  User guide update

### CODE REVIEW

File: src/frontends/tensorflow/src/frontend.cpp [hunk 1]
```diff
@@ -474,6 +479,13 @@ std::shared_ptr<ov::Model> FrontEnd::decode(const ov::frontend::InputModel::Ptr&
 }
 
 void FrontEnd::convert(const std::shared_ptr<ov::Model>& partiallyConverted) const {
+    std::cout << "convert part m" << std::endl;
```
Comment (rkazants) [inline src/frontends/tensorflow/src/frontend.cpp:482]:
```suggestion
```

**Review by rkazants — COMMENTED** [review]
can you please add tests that check input and output tensors names (or extend existing tests) to check:
1. TF1 from memory and disk will contain tensor names with `:`
2. TF2 from memory and disk will not contain `:`

Or just point it out that it already exists and where:)

**Comment by popovaan** [comment]
> can you please add tests that check input and output tensors names (or extend existing tests) to check:
> 
> 1. TF1 from memory and disk will contain tensor names with `:`
> 2. TF2 from memory and disk will not contain `:`
> 
> Or just point it out that it already exists and where:)

Added these tests to OVC python API tests: [tests/layer_tests/ovc_python_api_tests/test_tf.py](https://github.com/openvinotoolkit/openvino/pull/21075/files#diff-de36de51e9e096fd65ee6aead7ff94291298862ec3f84f43d53add035d490feb)

Please take a look.

File: src/frontends/tensorflow/src/frontend.cpp [hunk 2]
```diff
@@ -474,6 +479,12 @@ std::shared_ptr<ov::Model> FrontEnd::decode(const ov::frontend::InputModel::Ptr&
 }
 
 void FrontEnd::convert(const std::shared_ptr<ov::Model>& partiallyConverted) const {
+    std::map<std::string, std::unordered_set<std::string>> names;
+
+    for (const auto& result : partiallyConverted->get_results()) {
+        names[result->get_friendly_name()] = result->get_output_tensor(0).get_names();
+    }
```
Comment (rkazants) [inline src/frontends/tensorflow/src/frontend.cpp:486]: why do we need this? Looks like wa. Below is the same comment

File: src/frontends/tensorflow/src/input_model.cpp [hunk 3]
```diff
@@ -501,7 +508,14 @@ std::shared_ptr<InputModel> InputModel::InputModelTFImpl::get_body_input_model(
...
     }
-    return std::make_shared<InputModel>(body_graph_iterator, m_telemetry);
+    return std::make_shared<InputModel>(body_graph_iterator,
+                                        m_telemetry,
+                                        std::shared_ptr<ov::frontend::tensorflow::VariablesIndex>{},
+                                        nullptr,
+                                        nullptr,
+                                        nullptr,
+                                        false,
+                                        m_graph_iterator->tensor_names_need_indices());
```
Comment (rkazants) [inline src/frontends/tensorflow/src/input_model.cpp:518]: nit: I think let us leave it by default. For body graphs it is not critical at all.

File: src/frontends/tensorflow/src/frontend.cpp (hunk 2 above)
Comment (popovaan) [reply src/frontends/tensorflow/src/frontend.cpp:486]: This is needed to save output names from ov.Model() before decoding and restore them after work of translator's.  Otherwise translator's rewrite them in 490 line. I think it should be fixed if we align inner names in set_node_name().

Comment (rkazants) [reply src/frontends/tensorflow/src/frontend.cpp:486]: yes, let us do it `set_node_name`

File: src/frontends/tensorflow/src/input_model.cpp (hunk 3 above)
Comment (popovaan) [reply src/frontends/tensorflow/src/input_model.cpp]: Done.

File: src/frontends/tensorflow/src/frontend.cpp (hunk 2 above)
Comment (popovaan) [reply src/frontends/tensorflow/src/frontend.cpp]: As discussed offline if we fix set_node_name, NMS stops to work, where wrong tensor names are set for some reason. I suggest to fix it in separate PR.

Regurding WA in convert(), I removed it as agreed and corrected test that compares convert and decode+convert paths: https://github.com/openvinotoolkit/openvino/pull/21075/files#diff-16b422a3f3ea0023255bea536b7fb6ae89da562ebc7057c871f9fba35711761b

File: src/frontends/tests/frontend/shared/src/convert_model.cpp [hunk 4]
```diff
@@ -78,6 +78,7 @@ TEST_P(FrontEndConvertModelTest, test_decode_convert_equal_convert) {
     if (m_frontEnd->get_name() != "tf" && m_frontEnd->get_name() != "tflite") {
         func_comparator.enable(FunctionsComparator::NAMES);
     }
+    func_comparator.disable(FunctionsComparator::TENSOR_NAMES);
```
[the remark this answers is not in the document]
Comment (rkazants) [reply src/frontends/tests/frontend/shared/src/convert_model.cpp:81]: And make a note about it. BTW, in case of `decode` what tensor names will we have?

File: src/frontends/common/src/input_model.cpp [hunk 5]
```diff
@@ -154,3 +154,11 @@ void InputModel::set_tensor_partial_value(const Place::Ptr& place, const void* m
...
     FRONTEND_CALL_STATEMENT("set_tensor_partial_value", m_actual->set_tensor_partial_value(place, min_value, max_value))
 }
+
+bool InputModel::tensor_names_need_indices() const {
+    if (!m_actual) {
+        return false;
+    }
+    FRONT_END_CHECK_IMPLEMENTED(m_actual, tensor_names_need_indices);
+    FRONTEND_RETURN_STATEMENT("tensor_names_need_indices", m_actual->tensor_names_need_indices())
+}
```
Comment (rkazants) [inline src/frontends/common/src/input_model.cpp:164]: please remove. this method is not needed in the common frontend. This should truly TF specific method

File: src/frontends/tensorflow/src/translate_session.cpp [hunk 6]
```diff
@@ -570,6 +602,10 @@ void TranslateSession::translate_graph(const ov::frontend::InputModel::Ptr& inpu
                                             operation_name + "node specified as custom output does not exist");
                 auto result_node = std::make_shared<ov::opset8::Result>(node_outputs[port_index]);
                 result_node->set_friendly_name(model_output_name);
+                if (set_names_with_indices) {
+                    result_node->get_output_tensor(0).set_names(
+                        {make_output_tensor_name(result_node, operation_name, port_index)});
+                }
```
Comment (rkazants) [inline src/frontends/tensorflow/src/translate_session.cpp:608]: I am not sure that we really need `make_output_tensor_name`. I think we should modify `adjust_saved_model_names` and `set_node_name`.

File: src/frontends/tests/frontend/shared/src/convert_model.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (popovaan) [reply src/frontends/tests/frontend/shared/src/convert_model.cpp:81]: > BTW, in case of decode what tensor names will we have?

We will have `{<name>:<port>, <name>}`

File: src/frontends/common/src/input_model.cpp (hunk 5 above)
Comment (popovaan) [reply src/frontends/common/src/input_model.cpp]: Done.

File: src/frontends/tests/frontend/shared/src/convert_model.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (popovaan) [reply src/frontends/tests/frontend/shared/src/convert_model.cpp]: Done. In case of `decode` model will have both "out_name:0" and "out_name".

File: src/frontends/tensorflow/src/translate_session.cpp (hunk 6 above)
Comment (popovaan) [reply src/frontends/tensorflow/src/translate_session.cpp]: Moved alignment to `adjust_saved_model_names()`. I kept `make_output_tensor_name()`, as it contains the logic of making a resulting name with ":0". Please check.

File: src/frontends/tensorflow/src/frontend.cpp (hunk 2 above)
Comment (popovaan) [reply src/frontends/tensorflow/src/frontend.cpp]: The problem with NMS problem was because outputs of NMS have "name:1", "name:2"... indices in op names. And `set_node_name()` added additional :0 which resulted in "name:1:0", "name:2:0". The problem was fixed when I corrected `set_node_name()` to not add :0 for names with indices.

But there is another problem with `set_node_name()`. This method sets names for model `Parameter`'s. This leads to problem in MO preprocessing. MO preprocessing tries to find model input names among `Place`'s, which are set by operation names: 
https://github.com/openvinotoolkit/openvino/blob/c6258294fea969d2908efe841df068bdb98d2b88/src/frontends/tensorflow/src/input_model.cpp#L224C14-L224C14

So looks like we need to create `Place`'s in `InputModel` with tensor names to make preprocesssing work.

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 7]
```diff
@@ -39,9 +39,27 @@ def get_enabled_and_disabled_transforms():
...
 
-        if not any([item.is_equal(node) for item in model_inputs_or_outputs]):
+        node_found = False
+        for item in model_inputs_or_outputs:
+            if is_input:
+                name = new_node['input_name']
+                if name in item.get_names():
+                    node_found = True
+                    break
+                elif name.endswith(":0") and name[0:-2] in item.get_names():
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py]: Currently it won't work, due to naming of Places in InputModel. I created a separate ticket to cover this case: 129464

File: src/frontends/tensorflow/src/translate_session.cpp [hunk 8]
```diff
@@ -708,25 +734,31 @@ void TranslateSession::translate_graph(const ov::frontend::InputModel::Ptr& inpu
...
 
-        // it makes sense to use set because Parameter and Results nodes may have the common tensor
-        std::set<ov::Output<ov::Node>> ov_tensors;
-        for (const auto& param : params) {
-            ov_tensors.insert(param->output(0));
-        }
-        for (const auto& result : results) {
-            ov_tensors.insert(result->input_value(0));
-        }
+    // it makes sense to use set because Parameter and Results nodes may have the common tensor
```
[the remark this answers is not in the document]
Comment (popovaan) [reply src/frontends/tensorflow/src/translate_session.cpp]: Removed this code.

File: src/frontends/tensorflow/src/translate_session.cpp [hunk 9]
```diff
@@ -708,25 +734,31 @@ void TranslateSession::translate_graph(const ov::frontend::InputModel::Ptr& inpu
...
-        // it iterates through these tensors and adjusts their names
-        // by remaining only user specific names or mark as unused tensor (produced by TensorFlow)
-        for (auto ov_tensor : ov_tensors) {
-            adjust_saved_model_names(ov_tensor, saved_model_inputs, saved_model_outputs);
-        }
+    // it iterates through these tensors and adjusts their names
+    // by remaining only user specific names or mark as unused tensor (produced by TensorFlow)
+
+    auto tf_input_model = std::dynamic_pointer_cast<ov::frontend::tensorflow::InputModel>(input_model);
+    FRONT_END_GENERAL_CHECK(tf_input_model, "Got unexpected InputModel class");
```
[the remark this answers is not in the document]
Comment (popovaan) [reply src/frontends/tensorflow/src/translate_session.cpp]: Removed this code.

File: src/frontends/tensorflow/src/translate_session.cpp [hunk 10]
```diff
@@ -114,6 +136,10 @@ void adjust_saved_model_names(ov::Output<ov::Node>& ov_output,
         } else {
             signature_passed = false;
         }
+        if (tensor_names_need_indices) {
+            auto tensor_name = ov_output.get_names().size() > 0 ? ov_output.get_any_name() : "";
```
[the remark this answers is not in the document]
Comment (popovaan) [reply src/frontends/tensorflow/src/translate_session.cpp]: Removed this code.

File: src/frontends/tensorflow/src/translate_session.cpp [hunk 11]
```diff
@@ -53,14 +54,35 @@ std::vector<T> reorder_ops_by_names(const std::vector<std::string>& names, const
     return resulted_ops;
 };
 
+std::string make_output_tensor_name(std::string tensor_name, std::string node_name) {
```
[the remark this answers is not in the document]
Comment (popovaan) [reply src/frontends/tensorflow/src/translate_session.cpp]: Removed this code.

File: src/frontends/tensorflow/src/translate_session.cpp [hunk 12]
```diff
@@ -53,14 +54,35 @@ std::vector<T> reorder_ops_by_names(const std::vector<std::string>& names, const
...
+                                                              port_index_node,
+                                                              port_type);
+
+    if (op_name_from_res_node == operation_name && tensor_name != "") {
+        return operation_name + ":" + std::to_string(port_index);
+    } else {
+        return op_name_from_res_node + ":" + std::to_string(port_index_node);
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (popovaan) [reply src/frontends/tensorflow/src/translate_session.cpp]: Removed this code.

File: tests/layer_tests/ovc_python_api_tests/test_tf.py [hunk 13]
```diff
@@ -49,7 +53,7 @@ def create_keras_model(temp_dir):
     x1 = tf.keras.Input(shape=input_shape, name=input_names[0])
     x2 = tf.keras.Input(shape=input_shape, name=input_names[1])
     y = tf.nn.sigmoid(tf.nn.relu(x1 + x2))
-    keras_net = tf.keras.Model(inputs=[x1, x2], outputs=[y])
+    keras_net = tf.keras.Model(inputs=[x1, x2], outputs=[{"output": y}])
```
Comment (rkazants) [inline tests/layer_tests/ovc_python_api_tests/test_tf.py:56]: why is this changed? We should also support output list.

File: tests/layer_tests/ovc_python_api_tests/test_tf.py [hunk 14]
```diff
@@ -1045,4 +1049,110 @@ def test_mo_convert_tf_model(self, params, ie_device, precision, ir_version,
...
+        from openvino import convert_model
+
+        model, _, _ = create_tf_graph_def(None)
+        path = save_to_pb(model, tmp_dir)
+
+        ov_model = convert_model(path)
+        out_tensors = ov_model.outputs[0].get_names()
+
+        assert len(out_tensors) == 1
+        assert list(out_tensors)[0].endswith(":0")
```
Comment (rkazants) [inline tests/layer_tests/ovc_python_api_tests/test_tf.py:1072]: can we check that tensor name exactly equal to expected value instread of just containing `:0`?

**Review by rkazants — COMMENTED** [review]
Left comments for tests. Other looks good to me.

File: tests/layer_tests/ovc_python_api_tests/test_tf.py (hunk 13 above)
Comment (popovaan) [reply tests/layer_tests/ovc_python_api_tests/test_tf.py]: This change is not necessary. It was done only to reuse Keras model in new tests, which check output names and need named outputs.  
I removed this change and added another method that creates Keras model with named outputs.

File: tests/layer_tests/ovc_python_api_tests/test_tf.py (hunk 14 above)
Comment (popovaan) [reply tests/layer_tests/ovc_python_api_tests/test_tf.py]: Done.

File: tests/layer_tests/ovc_python_api_tests/test_tf.py [hunk 15]
```diff
@@ -1045,4 +1049,136 @@ def test_mo_convert_tf_model(self, params, ie_device, precision, ir_version,
...
+        tf.compat.v1.reset_default_graph()
+        from openvino.tools.ovc import convert_model
+
+        model, _, _ = create_tf_graph_def(None)
+
+        ov_model = convert_model(model)
+        out_tensors = ov_model.outputs[0].get_names()
+
+        assert len(out_tensors) == 1
+        assert list(out_tensors)[0].endswith("Sigmoid:0")
```
Comment (rkazants) [inline tests/layer_tests/ovc_python_api_tests/test_tf.py:1114]:
```suggestion
        assert list(out_tensors)[0] == "Sigmoid:0"
```

File: src/frontends/tensorflow/src/translate_session.cpp [hunk 16]
```diff
@@ -639,12 +639,11 @@ void TranslateSession::translate_graph(const ov::frontend::InputModel::Ptr& inpu
...
-                                        "Output port with index " + std::to_string(port_index) + " of " +
-                                            operation_name + "node specified as custom output does not exist");
-                auto result_node = std::make_shared<ov::opset8::Result>(node_outputs[port_index]);
-                result_node->set_friendly_name(model_output_name);
-                results.push_back(result_node);
+                if (node_outputs.size() > port_index) {
+                    auto result_node = std::make_shared<ov::opset8::Result>(node_outputs[port_index]);
+                    result_node->set_friendly_name(model_output_name);
+                    results.push_back(result_node);
+                }
```
Comment (rkazants) [inline src/frontends/tensorflow/src/translate_session.cpp:646]: why do we need this change?

Comment (popovaan) [reply src/frontends/tensorflow/src/translate_session.cpp:646]: It is needed as we set names with ":0" in `InputModel`. 
`model_output` with name that ends with ":0" is processed in `port_type == "out"` branch (otherwise it would be processed in `port_type == "none"` branch), with `port_index==0`. 
It has `node_outputs.size() == 0`, as it is output, so I removed the check, which checks `node_outputs.size() > port_index`. 

We still create new `Result` here if it is not terminating node, but allow places with ":0" to not have any outputs.

**Review by rkazants — APPROVED** [review]
